File: py/aoc2024/day11.py
"""
Day 11: Plutonian Pebbles
"""

from collections import Counter


def part1(data: str) -> int:
    return _solve(data, 25)


def part2(data: str) -> int:
    return _solve(data, 75)


def _solve(data: str, n: int) -> int:
    """
    >>> _solve("0 1 10 99 999", 1)
    7
    >>> _solve("125 17", 6)
    22
    >>> _solve("125 17", 25)
    55312
    """
    counts = Counter(map(int, data.split()))
    for i in range(n):
        next = Counter()
        for num, count in counts.items():
            if num == 0:
                next[1] += count
            else:
                string = str(num)
                if len(string) % 2 == 0:
                    next[int(string[: len(string) // 2])] += count
                    next[int(string[len(string) // 2 :])] += count
                else:
                    next[2024 * num] += count
        counts = next
    return counts.total()


parts = (part1, part2)

File: rs/src/day11.rs
use std::collections::BTreeMap;
use std::num::ParseIntError;

pub fn part1(data: &str) -> Result<u64, ParseIntError> {
    solve::<25>(data)
}

pub fn part2(data: &str) -> Result<u64, ParseIntError> {
    solve::<75>(data)
}

fn solve<const N: usize>(data: &str) -> Result<u64, ParseIntError> {
    Ok((0..N)
        .fold(
            {
                let mut counts = BTreeMap::<u64, u64>::new();
                for word in data.split_ascii_whitespace() {
                    *counts.entry(word.parse()?).or_default() += 1;
                }
                counts
            },
            |counts, _| {
                let mut next = BTreeMap::new();
                for (num, count) in counts {
                    if num == 0 {
                        *next.entry(1).or_default() += count;
                    } else {
                        let length = num.ilog10() + 1;
                        if length % 2 == 0 {
                            let divisor = 10u64.pow(length / 2);
                            *next.entry(num / divisor).or_default() += count;
                            *next.entry(num % divisor).or_default() += count;
                        } else {
                            *next.entry(2024 * num).or_default() += count;
                        }
                    }
                }
                next
            },
        )
        .into_values()
        .sum())
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE_1: &str = indoc! {"
        0 1 10 99 999
    "};

    static EXAMPLE_2: &str = indoc! {"
        125 17
    "};

    #[test]
    fn part1_examples() {
        assert_eq!(Ok(7), solve::<1>(EXAMPLE_1));
        assert_eq!(Ok(22), solve::<6>(EXAMPLE_2));
        assert_eq!(Ok(55312), solve::<25>(EXAMPLE_2));
    }
}
